libcst/metadata/expression_context_provider.py

```diff
@@ -53,82 +53,114 @@ class ExpressionContext(Enum):
     DEL = auto()
 
 
-NODE_TYPES_HAVE_CONTEXT = (
-    cst.Attribute,
-    cst.Subscript,
-    cst.StarredElement,
-    cst.Name,
-    cst.List,
-    cst.Tuple,
-)
-
-
-class ExpressionContextProvider(BatchableMetadataProvider[Optional[ExpressionContext]]):
-    """
-    Generate :class:`ExpressionContext` metadata for the following node types:
-    :class:`~libcst.Attribute`, :class:`~libcst.Subscript`,
-    :class:`~libcst.StarredElement` , :class:`~libcst.Name`, :class:`~libcst.List`,
-    :class:`~libcst.Tuple`. The provided context mimics the `expr_context
-    <https://docs.python.org/3/library/ast.html>`__ in ast.
-
-    Three context types :attr:`ExpressionContext.STORE`,
-    :attr:`ExpressionContext.LOAD` and :attr:`ExpressionContext.DEL` are provided.
-    """
-
-    def _check_type_and_set_metadata(
-        self, node: cst.CSTNode, context: Optional[ExpressionContext]
+class ExpressionContextVisitor(cst.CSTVisitor):
+    def __init__(
+        self, provider: "ExpressionContextProvider", context: ExpressionContext
     ) -> None:
-        if context is not None and isinstance(node, NODE_TYPES_HAVE_CONTEXT):
-            self.set_metadata(node, context)
+        self.provider = provider
+        self.context = context
 
-    def visit_Assign(self, node: cst.Assign) -> None:
+    def visit_Assign(self, node: cst.Assign) -> bool:
         for target in node.targets:
-            self._check_type_and_set_metadata(target.target, ExpressionContext.STORE)
+            target.visit(
+                ExpressionContextVisitor(self.provider, ExpressionContext.STORE)
+            )
+        node.value.visit(self)
+        return False
```

Comment:
I still really like the trick of stopping execution of the current visitor and swapping in a different one for the subtree. Its really elegant.